primitives/src/runtime_api.rs

```diff
@@ -119,38 +119,38 @@ use sp_std::{collections::btree_map::BTreeMap, prelude::*};
 
 sp_api::decl_runtime_apis! {
 	/// The API for querying the state of parachains on-chain.
-	#[api_version(2)]
+	#[api_version(3)]
```

Comment:
I think it's better to keep this at v2 until all networks are upgraded and mark the implementations as `v3` instead. Of course the question is do we care about this history or not.

In both cases we need to test this on Versi - new client with old runtime and vice versa.

Reply:
Or I am missing something and we want to eradicate v2 completely with this PR?

Reply:
Well, we don't want, but it seems we don't have any other choice. As we're changing `SessionInfo`, which is now v2, and we need to bump its version to v3 still supporting v2 via `#[changed_in(3)]`, we have to bump everything to v3. Let me know if you see better options.

Reply:
I was planning a test on Versi as the following: deploy new nodes with an old runtime, then perform the runtime upgrade (meanwhile testing the migration) and keep burning in, thus testing new nodes with the new runtime.

Old nodes are not supposed to work correctly with the new runtime. So we're deploying in two stages: first, release the new node, wait for the majority to upgrade, and then release the new runtime.

Reply:
I think it's better to leave the base version to 2 here and tag the new `session_info` function with v3:
```
#[api_version(3)]
fn session_info(index: sp_staking::SessionIndex) -> Option<v3::SessionInfo>;
/// Old method to fetch v2 session info.
#[changed_in(3)]
fn session_info(index: sp_staking::SessionIndex) -> Option<v2::SessionInfo>;
```
Then we can selectively enable v2 and v3 on different networks. E.g. rococo (which is versi) will implement v3, while polkadot and kusama can remain on 2. This way we won't force ourselves to move everything to v3 and we can take as much time for tests as we wish.

What do you think about this approach?

Reply:
But you cannot specify `#[changed_in(3)]` for anything if your base API version is 2. It wasn't clear to me, too, when I first met this problem. But after I did some research on how it works, I found that behavior makes perfect sense.

`#[changed_in(x)]` for `function()` generates implicit method `function_before_version_x()`. Procedural macros expand its implementation into a wasm call, which requests runtime to execute `function` but interprets and decodes its result into the previous structure version. That's how it looks like for `session_info()` and its version with `#[changed_in(3)]` for example:

```rust
        fn session_info(
            &self,
            __runtime_api_at_param__: &self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::BlockId<
                Block,
            >,
            index: sp_staking::SessionIndex,
        ) -> std::result::Result<
            Option<v3::SessionInfo>,
            self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::ApiError,
        > {
            let __runtime_api_impl_params_encoded__ = self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::Encode::encode(
                &(&index),
            );
            <Self as ParachainHost<
                _,
                _,
                _,
            >>::__runtime_api_internal_call_api_at(
                    self,
                    __runtime_api_at_param__,
                    self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::ExecutionContext::OffchainCall(
                        None,
                    ),
                    __runtime_api_impl_params_encoded__,
                    &(|version| { "ParachainHost_session_info" }),
                )
                .and_then(|r| std::result::Result::map_err(
                    <Option<
                        v3::SessionInfo,
                    > as self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::Decode>::decode(
                        &mut &r[..],
                    ),
                    |err| self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::ApiError::FailedToDecodeReturnValue {
                        function: "ParachainHost_session_info",
                        error: err,
                    },
                ))
        }

        #[deprecated]
        fn session_info_before_version_3(
            &self,
            __runtime_api_at_param__: &self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::BlockId<
                Block,
            >,
            index: sp_staking::SessionIndex,
        ) -> std::result::Result<
            Option<v2::SessionInfo>,
            self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::ApiError,
        > {
            let __runtime_api_impl_params_encoded__ = self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::Encode::encode(
                &(&index),
            );
            <Self as ParachainHost<
                _,
                _,
                _,
            >>::__runtime_api_internal_call_api_at(
                    self,
                    __runtime_api_at_param__,
                    self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::ExecutionContext::OffchainCall(
                        None,
                    ),
                    __runtime_api_impl_params_encoded__,
                    &(|version| { "ParachainHost_session_info" }),
                )
                .and_then(|r| std::result::Result::map_err(
                    <Option<
                        v2::SessionInfo,
                    > as self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::Decode>::decode(
                        &mut &r[..],
                    ),
                    |err| self::sp_api_hidden_includes_DECL_RUNTIME_APIS::sp_api::ApiError::FailedToDecodeReturnValue {
                        function: "ParachainHost_session_info",
                        error: err,
                    },
                ))
        }
```

So the base version of `session_info()` becomes a plain call returning a primitive of the same version as the base API version, and `#[changed_in(3)] session_info()` becomes an implicit call returning `v2` version.

And if you have `#[changed_in(x)]` with `x > base_api_version`, it cannot be resolved at all. You cannot resolve the same call to return two different primitive versions.

That's why having `#[changed_in(x)]` where `x > base_api_version` [is explicitly prohibited](https://github.com/paritytech/substrate/blob/2bfc1dd66ef32cf8beb90007dfb544a9d28f1b2f/primitives/api/proc-macro/src/decl_runtime_apis.rs#L463).

Reply:
Okay, this is definitely a valid point :) I think this is an invariant which I haven't covered when implementing the runtime api versioning.

Let me think about it a bit more and get back to you (tomorrow). Fixing this in substrate won't be hard but I want to think about what side effects it might have.

Reply:
If course the fix in substrate IS hard. We'll do it your way.

We'll implement a proper support for such cases because it sounds like something we might need often. The substrate issue discussing the details is  https://github.com/paritytech/substrate/issues/13138